submissions/dikun/frogger-game/app.js

```diff
@@ -0,0 +1,128 @@
+"use srtict";
+
+const CELL = {
+	width: 100,
+	height: 85
+};
+
+const GAME_FIELD = {
+	top: 0,
+	right: CELL.width * 5,
+	bottom: CELL.height * 5,
+	left: 0
+};
+
+const PLAYER_START = {
+	x: CELL.width * 2.02,
+	y: CELL.height * 4.7
+};
+
+const ENEMY_SPEED = {
+	min: 90,
+	max: 250 
+}; 
+
+const SPRITE = {
+	enemy: 'images/enemy-bug.png',
+	player: 'images/char-boy.png'
+};
+
+const speed = () => ENEMY_SPEED.min + Math.floor(Math.random() * ENEMY_SPEED.max);
+const OVERLAP_AMOUNT = 70;
+const BORDER = 20;
+
+const LINE = {
+	1: CELL.height - BORDER,
+	2: CELL.height * 2 - BORDER,
+	3: CELL.height * 3- BORDER
+};
+
+const Enemy = function (x, y, speed, player) {
+	this.x = x;
+	this.y = y;
+	this.speed = speed();
+	this.player = player;
+	this.sprite = SPRITE.enemy;	
+};
+
+Enemy.prototype.checkColission = function () {
+	if(this.player.x + OVERLAP_AMOUNT > this.x &&
+	this.player.x < this.x + OVERLAP_AMOUNT &&
+	this.player.y + OVERLAP_AMOUNT > this.y &&
+	this.player.y < this.y + OVERLAP_AMOUNT) {
+	this.player.x = PLAYER_START.x;
+	this.player.y = PLAYER_START.y;
+	}
+};
+
+Enemy.prototype.update = function (dt) {
+	if (this.x > GAME_FIELD.right) {
+		this.x -= GAME_FIELD.right + CELL.width;
+		this.speed = speed();
+		}
+	this.x += this.speed * dt;
+	this.checkColission();
+};
+
+Enemy.prototype.render = function () {
+	ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+const Player = function () {
+	this.x = PLAYER_START.x;
+	this.y = PLAYER_START.y;
+	this.sprite = SPRITE.player;	
+};
+
+Player.prototype.update = function (dt) {};
+
+Player.prototype.handleInput = function (keyPress) {
+	if (keyPress == 'left' && this.x > GAME_FIELD.left + CELL.width) {
+		this.x -= CELL.width;
+		}
+	if (keyPress == 'right' && this.x < GAME_FIELD.right - CELL.width) {
+		this.x += CELL.width;
+	}
+	if (keyPress == 'up' && this.y > GAME_FIELD.top) {
+		this.y -= CELL.height;
+	}
+	if (keyPress == 'down' && this.y < GAME_FIELD.bottom - CELL.height) {
+		this.y += CELL.height;
+	}	
+	else if (this.y  < GAME_FIELD.top) {			
+		this.x = PLAYER_START.x;
+		this.y = PLAYER_START.y;	
+		alert('You win!');	
+	}		
+};
+
+Player.prototype.render = function () {
+	ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+const player = new Player(PLAYER_START.x, PLAYER_START.y);
+
+const allEnemies = [];
+
+function createEnemyPerLine (line, numberOfEnemies) {	
+	for (let i = 0; i < numberOfEnemies; i++) {		
+		allEnemies.push(new Enemy((GAME_FIELD.left - CELL.width), LINE[line], speed, player));		
+	}
+}
+
+createEnemyPerLine(1, 1);
+createEnemyPerLine(2, 2);
+createEnemyPerLine(3, 1);
```

Comment:
Mapping was much cleaner and readable.
Now we have longer code and also repetition.
What if I want to add 20 more rows? How can I avoid adding 20 more lines of code?
Take a step back and think how could you optimize `[CELL.height - BORDER, CELL.height*2 - BORDER, CELL.height*3 - BORDER]`. There is a pattern there. You must see it better if the line of code modified as follows: `[CELL.height*1 - BORDER, CELL.height*2 - BORDER, CELL.height*3 - BORDER]`.

Think of chained mapping also look into what parameters a callback for `map` receives and you could leverage.

Examples:

```
[1, 5, 7, 3].map(heightMeters => heightMeters * 100).map((heightCentimeters, numberOfWalls) => buildHouseFragment(heightCentimeters, numberOfWalls + 1));
```

Of course the above is optimizable. We can just multiply source data by 100 when calling `buildHouseFragment`. This is specifically to make you start thinking of what numbers like 1, 2, 3 in your original fragment mean.

Reply:
Now I can **read** that initial arrays contains lines. That exaplains a lot to me as a peer developer and now I clearly understand what to do if I want to add yet another enemy to run across line 2 for example.
Although I'd suggest rename `line` as `rowNumber` or `laneNumber` or `trackNumber`

Reply:
I spent a few hours trying to understand how I can add a new bug in line 2. OMG. Thank you a lot for all your remarks. Hope it is fixed now too.